New Feature - ASHRAE 90.1 SZVAV control for Packaged Terminal Units.

rraustad opened the pull request:
Pull request overview
---------------------
ASHRAE Standard 90.1 requires low speed air flow when zone loads are 50% of design. ZoneHVAC:PackagedTerminal units now conforms to this specification. As zone loads increase, the fan speed increases while maintaining a user specified outlet air temperature. When fan speed reaches maximum speed, outlet air temperature is allowed to increase to maintain zone thermostat set point temperatures.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

rraustad committed:
NFP for SZVAV for Packaged Terminal Units.

rraustad committed:
Add design document to end of NFP

rraustad commented:
What do I need to do to pass the model array into a function? I need to be able to do this for fan coils, unitary systems and packaged terminal units. The array can't be local to the function, it has to be a reference or other so the SZVAV model can adjust variables as needed (e.g., UnitarySystem(x).FanPartLoadRatio = 0.0;)

    auto & SZVAVModel( UnitarySystem( UnitarySysNum ) );

    CalcSZVAVModel( SZVAVModel, CoolingLoad, HeatingLoad );

    void
    CalcSZVAVModel(
        Array1D< Real64 > & SZVAVModel,
        bool const & CoolingLoad,
        bool const & HeatingLoad
    );

    1>C:\Users\RRaustad\Documents\GitHub\EnergyPlus\src\EnergyPlus\HVACUnitarySystem.cc(7129): error C2664: 'void EnergyPlus::HVACUnitarySystem::CalcSZVAVModel(ObjexxFCL::Array1D<double>,const bool &,const bool &)': cannot convert argument 1 from 'EnergyPlus::HVACUnitarySystem::UnitarySystemData' to 'ObjexxFCL::Array1D<double>'

jasondegraw commented:
@rraustad I am not sure I understand completely what you want to do, but I am pretty sure that an array is not the best choice. Wouldn't a struct be a better choice to contain the information you need?

rraustad commented:
Jason, I'll try to explain. In the SZVAV model, calculations are made and stored in the member variables (e.g., UnitarySystem.FanPartLoadRatio). Those variables are used in other functions that are called from the SZVAV model. So what I need is to be able to act on the model member variables from within the SZVAV model. I'd rather not change the code so that these values are passed to functions instead of being available directly from the member array.

The method below creates a reference to a member array. I can use SZVAVModel.FanPartLoadRatio = 0.0; and this does what I need. However, I can't pass that into a function, or don't know how.

auto & SZVAVModel( UnitarySystem( UnitarySysNum ) );

How this (or other idea) gets done is the question here. Pass in an array, pass in a pointer, something else, etc. What I think is the main problem here is that I want to call this new function from FanCoilUnits, HVACUnitarySystem and PackagedTerminalHeatPump and in those models the member arrays are of different size and shape. If this isn't resolved, I have to write duplicate code in each of the equipment models.

Thinking about what you said with struct, each of the 3 models will need to hold the same information to be used within the SZVAV model (e.g., PLR, CoolCoilWaterInletNode, CoolAirMassFlowRate, etc.). So if that information can be organized, and that information can be passed to or used within the SZVAV model so that those variables are acted upon, then that might solve this.

If you'd rather we can discuss this briefly on a cc. Thanks.

Myoldmopar commented on design/FY2017/NFP-Single-Zone VAV Fan Control for PackagedTerminalUnits.md:
```diff
@@ -0,0 +1,510 @@
+
+The team will apply the previously developed SZVAV modeling methodology to the zone packaged terminal unit objects.
+
+The SZVAV model is now included in fan coil units and unitary systems. Regarding future maintenance it would be better to provide a common function for use by all equipment models. Since the SZVAV model uses specific array variables based on equipment type, a method to pass the necessary information via a function call is warranted. For example, a call to a common routine could pass all necessary ...
```
If we want to pursue the ultimate right way to do this, it's going to involve extracting parts of the individual data structures into a common base class.  If we can find what parts of each component are common, then we can write functions that accept any variation of this base class and can perform operations easily.  This will reduce the burden of adding `if` blocks inside the functions that are supposed to be generic.  I'd be happy to chat more about that aspect, but I won't just dump it all onto you in this comment.

Myoldmopar commented on design/FY2017/NFP-Single-Zone VAV Fan Control for PackagedTerminalUnits.md:
```diff
@@ -0,0 +1,510 @@
+        \object-list DesignSpecificationZoneHVACSizingName
+
+
+    New Inputs:
```
Glad these are at the end, and all have meaningful defaults, so no other IDD changes needed.  Unless...do you anticipate exposing this option in the HVACTemplate objects that create these components?

rraustad replied:
I'll let @mjwitte make the decision for HVACTemplate. Sounds like it would be worthwhile.

mjwitte replied:
@rraustad Always good to keep up with new options in HVACTemplate.

Myoldmopar commented on design/FY2017/NFP-Single-Zone VAV Fan Control for PackagedTerminalUnits.md:
```diff
@@ -0,0 +1,510 @@
+
+    auto & SZVAVModel( FanCoil[ FanCoilNum ] );
+    auto & SZVAVModel( UnitarySystem[ UnitarySystemNum ] );
+    auto & SZVAVModel( PTUnit[ PTUnitNum ] );
```
Hmmm, I think we'll need to chat about this approach.  I'm not sure this will work in the way you are wanting it to work, but 1) I may be missing something and 2) I think we can definitely find a way that _will_ work.

rraustad replied:
I agree.

Myoldmopar reviewed (commented):
This is a great addition @rraustad.  I made some comments.  I'd like to chat more about the use of polymorphism in your design.  I know tomorrow's (11/22) sizing call is cancelled, but maybe it's worth having a brief call next week so we don't have to wait until the next sizing call comes around.

Myoldmopar commented:
And I just saw your comments with @jasondegraw.  Indeed, that's the part that I didn't think would work as expected.  Let's chat next week if you're free and I think we can work out a plan; Monday?

rraustad commented:
@Myoldmopar can you offer any advice for how to be able to call the same function from multiple models while still using the model member array as a data source for read/write? See my description above. Could I use a pointer as a function argument? Can the member arrays be set up so that I have access to a subset of variables (i.e., only those needed by the SZVAV model), other ideas? I am already resolved to having about 10 or so variables of the same name in each of the 3 models (e.g., UnitarySystem().FanPartLoadRatio, FanCoil().FanPartLoadRatio and PTUnit().FanPartLoadRatio plus several others of the same name). So there will be a subset of variables with the same name across these 3 models. If I can read/write from/to a specific variable (e.g., FanPartLoadRatio) within the SZVAV model and have that act on the correct model then I could get this to work. 

Remember that I am trying to write a generic function (CalcSZVAVModel) to be used by all models. If I call function CalcUnitarySystem from within the new CalcSZVAVModel function, then CalcUnitarySystem will be using UnitarySystem().FanPartLoadRatio and if I call CalcFanCoilUnit from within the new function then CalcFanCoilUnit will be using FanCoil().FanPartLoadRatio. I hope I have explained this well enough to get a little advice.

I just saw your post and think a cc is worth the time. Monday any time would work for me.

Myoldmopar commented:
Monday we'll chat.  But yes, the way to do what you want to do is to create a single base class that defines the common variables.  Let's work through a quick example, with an understanding that I'm typing, not compiling, so consider it pseudo-code.

We have two coils to start:
```
struct CoilA {
  std::string name;
  Real64 flowRate;
  int specialCoilAParameter;
};
struct CoilB {
  std::string name;
  Real64 flowRate;
  int OTHERCoilBParameter;
};
```

Note there is some common variables there, namely the name and the flowRate.  Now, we want to write a function that will accept a coil and do something to the flow rate.  However, we can't pass a variable that is **either** `CoilA` **or** `CoilB`, because they are completely unrelated.  However, let's create a base struct, and make each coil inherit from it.

```
struct BaseCoil {
  std::string name;
  Real64 flowRate;
};
struct CoilA : BaseCoil {
  int specialCoilAParameter;
};
struct CoilB : BaseCoil {
  int OTHERCoilBParameter;
};
```

Note how the coils no longer need to define the common variables, just the ones special to them.  Even with just this simple change, we've already reduced our amount of code 👍 .  But now we want to pass in an instance of a coil to a function.  It would be nice if we could just:

```
void setFlowRate( BaseCoil thisCoil ) {
  thisCoil.flowRate = 3.14;
}
```

But, afaik, if we try to pass in a CoilA instance, it won't run properly, simply because CoilA isn't BaseCoil.  But they **are** related, so if we make a small adjustment, we can pass in a derived class, like so:

```
void myFunction( BaseCoil * thisCoil ) {
  thisCoil.flowRate = 3.14;
}
```

Now we are passing a pointer, not a whole instance.  And the compiler can do the vtable lookup magic to handle polymorphism properly, and all will be happy.  We call this function like so:

```
int myWorker() {
  CoilA my_coil_a;
  myFunction( &my_coil_a ); // and boom, this will adjust the flow rate
}
```

OK, so that should work-ish.  But, anytime we start using `*` or `&` in C++, we need to question whether we are making the right decision.  The main reason we don't like using raw pointers is because we have to manage the lifetime of the object.  In this case, we are creating a `CoilA` instance inside `myWorker`, on the stack as a local variable, so we aren't responsible for cleaning up any pointers.  The pointer operation is simply passing the address of a regular variable into `myFunction`, which doesn't get us involved with object lifetime.  So we are pretty safe moving forward here.

Now, I'm going to do the right thing here, and tag @mbadams5, so that he can come and disagree with everything and show me I'm doing it all wrong.  🙃 😄 🙃

jasondegraw commented:
So, yeah, a base class would work @Myoldmopar, but... if I understand what @rraustad is saying, then a template function  would be a better and more general approach:

```
template <typename T> double doSomethingWithFanPartLoadRatio(T &object)
{
    object.fanPartLoadRatio *= 2;
    return 3.14*object.fanPartLoadRatio;
}
```

The pointer business is avoided with this. I'd personally like to avoid having a lot of functions that take pointers as arguments and rely on inheritance. If we're trying to write a generic function, I'd rather do it for real.

Two downsides to templates are (1) the cryptic compiler error messages you get when something isn't quite right and (2) longer compile times if you go overboard. But there's Google for the error messages, and we can be careful about introducing this sort of thing.

rraustad commented:
The template approach sounds promising since there are only 3 structure variables that need updating within CalcSZVAVModel: FanPartLoadRatio, CoolCoilWaterFlowRatio and HeatCoilWaterFlowRatio. Well, 3 in UnitarySystem. I see maybe 1 in fan coil and not sure how many yet in the PTUnit model. But it sounds like only a few regardless of system type. 

The other variables needed within the CalcSZVAVModel function are const and can be passed in using a temp array (like we do with Par), unless something can be done where I can directly access the member variables generically. So in this case a base class would allow generic access to member variables depending on where this function is called from?

Myoldmopar commented:
I don't have a strong opinion on which way you go in this specific task, but I'll just say when we do the coil refactor and try to collapse multiple coils into one, we'll be extracting out base class variables and methods.

I know I mentioned pointers in my comment above, and it looks funny because in that exact case, we are trying to shove multiple shaped pegs into a standalone function.  But that's really only the first step.  If that approach gets pushed forward, the next step is taking that pointer-accepting-function and just dropping it into the base class itself, without any arguments.  There will still be pointers and polymorphism, but the management of pointers is pushed to a different level where the everyday development doesn't get bogged down with pointer arguments and blah :+1:, and no template issues to worry about either :+1:.

rraustad commented:
Your not having a strong opinion is better than my lack of opinion. These look very similar to me from my rudimentary understanding of C++. And since I don't know what I'm doing, I like the reference method better. But what about future upgrades, would this change be reverted to something else? That seems like a waste of effort.

So now I can "set" a member variable in different struct's (as long as they have the same name). But I still need access to other member variables for example coil inlet node number, coil air outlet node number, etc. Take this example:

    SetComponentFlowRate( mdot, FanCoil( FanCoilNum ).HotControlNode, FanCoil( FanCoilNum ).HotPlantOutletNode, FanCoil( FanCoilNum ).HWLoopNum, FanCoil( FanCoilNum ).HWLoopSide, FanCoil( FanCoilNum ).HWBranchNum, FanCoil( FanCoilNum ).HWCompNum );

and for UnitarySystem it would be:

    SetComponentFlowRate( mdot, UnitarySystem( UnitarySystemNum ).HotControlNode, UnitarySystem( UnitarySystemNum ).HotPlantOutletNode, UnitarySystem( UnitarySystemNum ).HWLoopNum, UnitarySystem( UnitarySystemNum ).HWLoopSide, UnitarySystem( UnitarySystemNum ).HWBranchNum, UnitarySystem( UnitarySystemNum ).HWCompNum );

Yes, these member variables can be passed in a temp array. Is that the best choice? Maybe for now the answer is yes?

rraustad commented:
Why couldn't I create a template for this new function where the reference to the member array is passed in and the type changes based on where this function is called? Wouldn't I then have access to the member variables?

jasondegraw commented:
I'm not sure what you mean by "member array", so I'm not sure I can answer. Let's have a call Monday, otherwise this thread could get rather long and confusing. @Myoldmopar you want to schedule?

rraustad commented:
e.g., I think of Name and NumOfSpeedCooling, etc., as a member variable. More clarity on Monday.

    struct DesignSpecMSHPData
    {
        // Members
        std::string Name; // Name of the design specification MSHP
        int NumOfSpeedCooling; // The number of speeds for cooling
        int NumOfSpeedHeating; // The number of speeds for heating
        Array1D< Real64 > CoolingVolFlowRatio; // The ratio of flow to max for this speed
        Array1D< Real64 > HeatingVolFlowRatio; // The ratio of flow to max for this speed
        bool SingleModeFlag; // If yes, single mode operation
        Real64 NoLoadAirFlowRateRatio; // fraction of supply air flow rate multiplier

        // Default Constructor
        DesignSpecMSHPData() :
            NumOfSpeedCooling( 0 ),
            NumOfSpeedHeating( 0 ),
            SingleModeFlag( false ),
            NoLoadAirFlowRateRatio( 1.0 )
            {}
    };

rraustad committed:
Move SZVAV code to new function

rraustad committed:
Update with #6406 changes

rraustad commented:
Changes for UnitarySystem where SZVAV code was moved to new function in General.cc. Results look the same as for V8.8 (small diff's are misleading).

![szvav_resultsofnewfunction](https://user-images.githubusercontent.com/5778156/33393467-4322d4ea-d50d-11e7-973f-db10ff42a79c.jpg)

rraustad committed:
Update Final NFP and add PTHP code. Still needs testing.

rraustad committed:
Add SZVAV control to PTAC.

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into SZVAV-for-PTUnit

Conflicts:
	src/EnergyPlus/HVACUnitarySystem.cc

rraustad commented:
@Myoldmopar I've never seen this one before. Both CIs report this same warning.

error: RPC failed; result=56, HTTP code = 200 fatal: The remote end hung up unexpectedly fatal: early EOF fatal: index-pack failed

rraustad commented:
@Myoldmopar @mbadams5 @jasondegraw @JasonGlazer @mjwitte I am a bit confused here. I see the HospitalLowEnergy test file failing for a reason I can figure out. I run this file in debug and it runs. In release it does not. So I added some display text messages to narrow this down. When I run in release mode the program stops at my new message "Initialize Simulation1" (right after the existing message "Initialize Simulation"). This points to:

    KickOffSimulation = true;

But when I run in debug mode, I can step right past this and STILL don't see my messages. ??? Since I can't figure this out I am thinking this is deeper than my knowledge this started happening when I added the second function = calcSZVAVModel. What I did was add a function in General.cc of calcSZVAVModel for HVACUnitarySystem, that seemed to work (did work as best I can tell). Then I added another identical function for PTUnit, with the same name, but different class, and things started getting weird. I just wanted to see no diff's before I proceeded. HospitalLowEnergy is failing (as are others). If one of you could help identify what is wrong I would be very grateful. I've tried several things with no success. I'm going to upload what I have at this point since I am stumped (and the hung up failure seems bogus) and want a picture of where I'm at.

Note here that my release version (Win64) stops at "Initialize Simulation1", which is the string I added. But does not show a message for "ResetEnvironmentCounter", which I also added, if I step through in debug mode, ??? It must be KickOffSimulation (doesn't seem likely)? I tried adding the class, the includes, etc. but nothing seems to work. If I step past this in debugger I still don't see my messages. ??? I've really tried everything I can think of. Any help would be appreciated. I'm on vacation so am in no hurry. BUT, if something falls down my chimney unexpectedly, those that provide the gift (S8 or 3D Vision) might prosper for years to come ("might", but probable), Merry Christmas everybody!


![simulation1message](https://user-images.githubusercontent.com/5778156/33915881-e1a52af8-df73-11e7-8535-076e65e4229a.jpg)

rraustad committed:
Debugging HospitalLowEnergy issue...???

rraustad commented:
I've made some progress in figuring this out. What threw me is that DisplayString doesn't display when KickOffSimulation=true. So instead, I've added cout statements and tracked this down to a specific PTHP that is causing the crash. Not sure why yet but I'm closer. It crashes when ControlZoneNum=31, just need to figure out why.

![debug](https://user-images.githubusercontent.com/5778156/33974301-be013d3e-e055-11e7-81d5-a4db97b18567.jpg)

rraustad committed:
Fix crash with HospitalLowEnergy and move SZVAV model code to new module.

rraustad commented on src/EnergyPlus/PackagedTerminalHeatPump.cc:
```diff
@@ -3519,7 +3522,7 @@ namespace PackagedTerminalHeatPump {
 
 			//   set fluid-side hardware limits
-			if ( PTUnit( PTUnitNum ).HeatCoilFluidInletNode > 0 ) {
+			if ( PTUnit( PTUnitNum ).HeatCoilFluidInletNode > 0 && PTUnit( PTUnitNum ).UnitType_Num != PTWSHPUnit ) {
```
This is what caused the crash in release mode. The WSHP water inlet node was not available and caused a crash when InitComponentNodes was called (I changed the name of this variable and this popped up). I think maybe that InitComponentNodes should also be called for WSHP coils but will address that later.

JasonGlazer commented:
It is always annoying when even the debugger doesn't work as expected and you have to revert to using cout to do some old fashioned debugging.

rraustad committed:
Correct issue with failing test files, organize SZVAV variables, unify SZVAV variable names in UnitarySys and PTUnit

rraustad commented on src/EnergyPlus/HVACUnitarySystem.hh:
```diff
@@ -489,23 +457,53 @@ namespace HVACUnitarySystem {
 		int FaultyCoilSATIndex;  // Index of the fault object corresponding to the coil
 		Real64 FaultyCoilSATOffset; // Coil SAT sensor offset
 		
+		// variables used in SZVAV model:
```
I moved all member variables used in SZVAV to here so that we know which ones are needed in the base class. Each of these variables have the same name in HVACUnitarySystem and PackagedTerminalHeatPump.

Myoldmopar replied:
:+1: !

rraustad committed:
Increase array size for Par to correct unit test failure

rraustad committed:
Updates to SZVAV model

rraustad committed:
Added IORef and EngRef documentation.

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into SZVAV-for-PTUnit

rraustad committed:
Address compile warnings

rraustad committed:
Revise SZVAV model for PTUnit. More compact now and should be faster.

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into SZVAV-for-PTUnit

rraustad committed:
Rework SZVAV model, only 3 steps now, and faster.

rraustad commented:
This should be near the final form. I had what I thought were issues and reworked the SZVAV model code completely. Now only 3 steps (instead of 9) and should be faster. The answer looks the same as before for UnitarySystem and also looks fairly good for PTUnit.

Results for UnitarySystem:

![testairloophvacunitarysystem](https://user-images.githubusercontent.com/5778156/35352935-79909060-013d-11e8-9d04-4eb3db01b3df.png)

Results for PTUnit:

![ptunit_szvav_fancontrol](https://user-images.githubusercontent.com/5778156/35352966-91a58cf0-013d-11e8-84f1-b8fa123f8f93.png)

rraustad commented:
There are 2 unused variables that need correcting. I'll await comments before correcting those.

    warning | src/EnergyPlus/SZVAVModel.cc:226:10 | unused variable 'SupHeaterLoad' [-Wunused-variable]
    warning | tst/EnergyPlus/unit/PackagedTerminalHeatPump.unit.cc:913:10 | unused variable 'PrimaryAirMassFlowRate' [-Wunused-variable]

nrel-bot-3 commented:
@rraustad @lgentile it has been 14 days since this pull request was last updated.

Myoldmopar commented on src/EnergyPlus/HVACUnitarySystem.cc:
```diff
@@ -5965,11 +5973,11 @@ namespace HVACUnitarySystem {
 						ShowContinueError( "When a draw through fan is specified, the fan outlet node name must be the same as the unitary system outlet node name." );
 						ShowContinueError( "...Fan outlet node name           = " + NodeID( FanOutletNode ) );
-						ShowContinueError( "...UnitarySystem outlet node name = " + NodeID( UnitarySystem( UnitarySysNum ).UnitarySystemOutletNodeNum ) );
+						ShowContinueError( "...UnitarySystem outlet node name = " + NodeID( UnitarySystem( UnitarySysNum ).AirOutNode ) );
```
Wow tons of these changes are just changing the variable being used in function calls and evaluations.

rraustad replied:
Yes, variable names were unified across models. Who would have thought we should have been doing this all along.

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -818,6 +818,7 @@ HVACSystemRootFindingAlgorithm,
        \key Bisection
        \key BisectionThenRegulaFalsi
        \key RegulaFalsiThenBisection
+       \key Alternation
```
Alternation.....sounds like an Alternative Rock Radio Station name: 104.3 The AlterNation!

Myoldmopar commented on doc/input-output-reference/src/overview/group-zone-forced-air-units.tex:
```diff
@@ -2615,6 +2628,19 @@ \subsubsection{Inputs}\label{inputs-9-026}
+
+When Capacity Control Method = SingleZoneVAV, enter the minimum air temperature limit for reduced fan speed in cooling mode. For SingleZoneVAV, this field's minimum supply air temperature is 0.0\(^{o}\)C. This field is autosizable and the default is autosize.
+
+\paragraph{Field: Maximum Supply Air Temperature in Heating Mode}\label{field-maximum-supply-air-temperature-in-heating-mode-3}
```
The docs look good.  Just a total aside here.  You don't *have* to label every object in the documentation.  Only if you are referring back to them at some other point in the documentation.  It doesn't hurt it to have them, except maybe extra memory and computation when building the docs, so go ahead and leave them, but just in the future you should know it's not mandatory.

rraustad replied:
Well, again, I don't know the rules, monkey see. There is rarely if any, I've never seen any, a reference, so why are these here in the first place?

Myoldmopar commented on src/EnergyPlus/SZVAVModel.hh:
```diff
@@ -0,0 +1,129 @@
+// EnergyPlus, Copyright (c) 1996-2017, The Board of Trustees of the University of Illinois,
```
The year needs to be updated to 2018 in this header file and the accompanying cc file.

rraustad replied:
Has it been that long since I finished this? Will do.

Myoldmopar commented on src/EnergyPlus/SZVAVModel.cc:
```diff
@@ -0,0 +1,712 @@
+		Par( 16 ) = 1.0; // iteration method, 1 = modulate coil capacity, 2 = modulate air flow rate
+		Par( 17 ) = double( CompressorONFlag ); // ** not used, gets rid of warning (unused variable) in PTUnit version of SZVAV
+
+		int InletNode = SZVAVModel.AirInNode;
```
Bonus for declaring variables right where they are used.  :+1:

Myoldmopar commented on src/EnergyPlus/SZVAVModel.cc:
```diff
@@ -0,0 +1,712 @@
+		}
+
+		if ( ( CoolingLoad && boundaryLoadMet < ZoneLoad ) || ( HeatingLoad && boundaryLoadMet > ZoneLoad ) ) { // in Region 1 of figure
+																												// Step 1: set min air flow and full coil capacity
```
What happened to this line?  I know your in Florida but no reason to make this comment so far East on my screen.  (Yeah yeah technically it's out west at my south facing desk)

rraustad replied:
My editor right justifies comment lines. I don't know why and it pisses me off. Will fix.

Myoldmopar reviewed (changes requested):
This seems pretty good.  The only absolute minimum change is the license year in the new SZVAV source files, because CI is complaining about it.  I had a few other comments, but most were just random notes.  If it turns out the license year is the only change, I can make that change and merge it into develop.

Myoldmopar commented:
The Linux CI failures are unrelated.

The things that need to be cleaned up include:
 - License Year in 2 files
 - Unused variables warning
 - One commented line is line 20 tabs over to the right

Otherwise I am good with this.

rraustad commented:
Thanks for the review @Myoldmopar. @mjwitte do you want to take a quick pass at this? No need to spend a lot of time, but just get a feel for this change. Common code was moved to a common location in preparation to create a single SZVAV function that could be used by UnitarySystem. PT Units and FanCoils..

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into SZVAV-for-PTUnit

rraustad committed:
New doc figure?

rraustad commented:
@mjwitte I'm confused. In epfilter.f90 I am trying to add SZVAV. I see for fan coils:

  CALL AddToObjStr('Maximum Flow Rate {m3/s}','autosize')

Yet the field name is:

ZoneHVAC:FourPipeFanCoil,
   N1 , \field Maximum Supply Air Flow Rate

How does this work?  Isn't this looking for a field name? or is it just setting a field name in the correct order? Does validateFieldNameInIDD check for correct field names in the IDD?

rraustad commented:
See epfilter.f90 line 23699

mjwitte commented:
Many of the field names are old. They just end up in the endline comments. It's all order dependent. Mike

-------- Original message --------
From: Richard Raustad <notifications@github.com> 
Date: 2/16/18  9:04 PM  (GMT-06:00) 
To: NREL/EnergyPlus <EnergyPlus@noreply.github.com> 
Cc: "Michael J. Witte" <mjwitte@gard.com>, Mention <mention@noreply.github.com> 
Subject: Re: [NREL/EnergyPlus] New Feature - ASHRAE 90.1 SZVAV control for
&nbsp; Packaged Terminal Units. (#6393) 

@mjwitte I'm confused. In epfilter.f90 I am trying to add SZVAV. I see for fan coils:
CALL AddToObjStr('Maximum Flow Rate {m3/s}','autosize')
Yet the field name is:
ZoneHVAC:FourPipeFanCoil,

N1 , \field Maximum Supply Air Flow Rate
How does this work?  Isn't this looking for a field name? or is it just setting a field name in the correct order? Does validateFieldNameInIDD check for correct field names in the IDD?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/NREL/EnergyPlus","title":"NREL/EnergyPlus","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/NREL/EnergyPlus"}},"updates":{"snippets":[{"icon":"PERSON","message":"@rraustad in #6393: @mjwitte I'm confused. In epfilter.f90 I am trying to add SZVAV. I see for fan coils:\r\n\r\n  CALL AddToObjStr('Maximum Flow Rate {m3/s}','autosize')\r\n\r\nYet the field name is:\r\n\r\nZoneHVAC:FourPipeFanCoil,\r\n   N1 , \\field Maximum Supply Air Flow Rate\r\n\r\nHow does this work?  Isn't this looking for a field name? or is it just setting a field name in the correct order? Does validateFieldNameInIDD check for correct field names in the IDD?"}],"action":{"name":"View Pull Request","url":"https://github.com/NREL/EnergyPlus/pull/6393#issuecomment-366411454"}}}

rraustad committed:
Merge branch 'SZVAV-for-PTUnit' of https://github.com/NREL/EnergyPlus into SZVAV-for-PTUnit

Conflicts:
	src/EnergyPlus/HVACUnitarySystem.cc
	src/EnergyPlus/PackagedTerminalHeatPump.cc
	src/EnergyPlus/SZVAVModel.cc
	src/EnergyPlus/SZVAVModel.hh

rraustad committed:
typo correction

Myoldmopar commented:
My review comments were addressed, so I'll mark my review approved, however, there is a build failure in the FORTRAN code.  I'll take a peek.

Myoldmopar commented on src/ExpandObjects/epfilter.f90:
```diff
@@ -23655,7 +23661,7 @@ SUBROUTINE CreateFanCoil
   CALL AddToObjStr('Outdoor Air Mixer Object Type', 'OutdoorAir:Mixer')
   CALL AddToObjFld('Outdoor Air Mixer Name', base + fczNameOff,' OA Mixing Box')
-  IF (capacityControlKind == capctrlConstFanVarFlow) THEN
+  IF (capacityControlKind == capctrlConstFanVarFlow,capctrlASHRAE90VariableFan) THEN
```
I don't think this works, right?

Myoldmopar committed:
Fix fortran IF condition

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into SZVAV-for-PTUnit

Myoldmopar commented:
OK, I fixed the Fortran build issue and merged develop in.  If CI is happy, I think this is ready.

Myoldmopar commented:
Ugh.  Build warnings for two unused variables and one expected diffs.  I am going to push up the unused variable fixes after I verify they build locally and then merge into develop.  We still have many branches to get in and CI is going really slow.

Myoldmopar committed:
Fix unused variables

rraustad commented on src/ExpandObjects/epfilter.f90:
```diff
@@ -23467,6 +23469,10 @@ SUBROUTINE CreateFanCoil
     capacityControlKind = capctrlVarFanConstFlow
+!  ELSEIF (SameString(FldVal(base +  fczCapCtrlTypeOff),'MultiSpeedFan')) THEN
+!    capacityControlKind = capctrlMultiSpeedFan
+  ELSEIF (SameString(FldVal(base +  fczCapCtrlTypeOff),'ASHRAE90VariableFan')) THEN
```
I thought I added a new key here in the idd. I don't see it now so HVAC template for fan coil will not be able to create the SZVAV model (yet). These changes shouldn't hurt anything but SZVAV won't work yet. Also, the SZVAV model for UnitarySystem and PackagedTerminalUnit will also need IDD changes since the HVAC template model for these will need a key to know when to create the SZVAV model. I think I'll wait on this since there is so much other work to do.

rraustad commented on src/ExpandObjects/epfilter.f90:
```diff
@@ -23691,6 +23697,7 @@ SUBROUTINE CreateFanCoil
     CALL AddToObjFld('Fan Efficiency', base + fczSupplyEfficiencyOff,'')
     CALL AddToObjFld('Pressure Rise {Pa}', base + fczSupplyPressureOff,'')
     CALL AddToObjStr('Maximum Flow Rate {m3/s}','autosize')
+	if ( capacityControlKind .EQ. capctrlASHRAE90VariableFan ) CALL AddToObjStr('Low Speed Supply Air Flow Ratio','autosize')
```
This is also wrong, and should have been in the fan coil unit not the fan. Again, won't hurt anything since capacityControlKind can never be capctrlASHRAE90VariableFan since that key is not in the idd. I'll correct this when I get a chance.

mjwitte replied:
OK on waiting with HVACTemplate support for the new control option.  Be sure to start a new branch/PR for that.
